### PR TITLE
Chef-13: Freeze property defaults

### PR DESCRIPTION
**Pull request by coderanger**
Been planned for a while now, and should prevent a bunch of bugs like the one we had in Mount's tests :)

### CODE REVIEW

File: lib/chef/property.rb [hunk 1]
```diff
@@ -110,6 +110,9 @@ def initialize(**options)
         raise ArgumentError, "Cannot specify both default and name_property/name_attribute together on property #{self}"
       end
 
+      # Freeze the default if it isn't a lazy value.
+      default.freeze unless default.is_a?(DelayedEvaluator)
```
Comment (lamont-granquist) [inline lib/chef/property.rb]: what about hashes and arrays?  pretty sure the yum_package default mutation bug was on a hash key.

**Comment by coderanger** [comment]
Travis failures look unrelated. Merging.

**Comment by lamont-granquist** [comment]
#4691 is the "reference bug" for why we do that.  the yum_provider flush_cache default mutation bug was awful.

this also already found errors in the tomcat cookbook.

**Comment by lamont-granquist** [comment]
we also discussed and could not design a way to provide deprecation warnings for this, since you'd have to give back some kind of decorator over the object which gave deprecation warnings if any mutator methods were called, which would be nearly impossible due to both ruby being bad about decorators (truthiness, hardwired =~ operator, === comparisons, etc) and knowing what the mutator methods were (i supposed you could catch the exception, unfreeze, replay the method, and refreeze or something insane like that -- but still you'd break all kinds of code.

**Pull request merged**